### DISCUSSION TITLE
Master volume?

### DISCUSSION THREAD

**Discussion by eLuSiVeMiTe** [probonopd/MiniDexed#157] [General]
Noticed a big increase in overall volume when I updated to a current build yesterday. 
Left me a bit deafened when testing with headphones.

Have I missed a master volume or is it not present?

**Comment by probonopd** [comment]
Volume is adjustable via the menu for each tone generator and via MIDI CC 10.

**Comment by probonopd** [comment]
> Noticed a big increase in overall volume when I updated to a current build yesterday.

Did you update from a version older than, say, a week?

**Reply by eLuSiVeMiTe** [reply]
Aye. Would have been the 17th I was running previously.

I get what your saying re the tone generators volumes (great for mixing )but every patch would need turning down everytime I switch bank/voice. A master volume to lower the total output to a suitable level across the board makes sense.

**Reply by probonopd** [reply]
If you set your Tone Generators (TGs) to MIDI channel "Omni" and use the volume slider/knob on your MIDI controller, the volume of all TGs gets changed.

Can you find out how the TX816 and TX802 deal with this?

**Reply by eLuSiVeMiTe** [reply]
Every multitimbral synth I've owned or used has a master output level and individual for each part. 
Not owned a DX or its siblings but I can have a nosey

**Reply by probonopd** [reply]
Should check with the TX816 and TX802 manuals, especially which CC messages are interpreted for the per-TG vs global master volume.

**Reply by bmillier** [reply]
I used a TX802 for many years before I sold it. As far as I remember, there were  volume controls for each T.Generator, which responded to volume CC messages on a per channel basis. I don't remember any master volume setting- either in software via MIDI command, or a volume pot on the front panel. Looking at the TX802 user manual, I don't see anything to contradict this in there either.
  I agree with craigyjp that it doesn't make sense to change the volume on all generators by having each generator set for MIDI OMNI  as you are then reducing it to a non multi-tembral instrument. On that topic, the default performance.ini file has TG1-4 set for MIDI OMNI. As a newcomer to MiniDexed, I didn't at first look at this file. So, if you start out by changing the voice on TG1, for  example, you get a mixture of the chosen voice, mixed in with the default voices of TG2-4. So, everything sounds weird, and doesn't change much when you select different voices for TG1. When I examined performance.ini, I figured out what was going on. Might I suggest that you set TG1 for OMNI, and the other TGs for MIDI channels 2,3,4 etc. (or disable them)  Then a newcomer won't get this mashup of voices when they try to change a TG1 voice - when they are starting from scratch.
   While one could add software to scale the data going to the DAC, for master volume, I would personally just use a stereo pot at the output of the DAC. You'd have to install new sockets after the pot, and use the DAC breakout board's output pins or socket to feed the pots. Basically, as craigyjp suggests.
   Question- the I2S data coming from the Raspi- is it 16, 20, 24 or 32-bit?  If you don't know, I'll connect up a logic analyser and figure it out that way. If it is only 16-bit, the advantages of the PCM5102 (up to 32-bit) are not as meaningful vs the 16-bit UDA1334 breakout. While I've ordered a PCM5102 breakout (from China) , if it never arrives I may just get the UDA1334 breakout from Mouser. 
  You mentioned the WIKI addition you made re the voice banks folder/names that I had suggested. I checked this out- Thanks that is exactly what I thought would be useful.
    Cheers

**Comment by craigyjp** [comment]
Yeah but they all receive the same notes and pgm changes too which defeats
the point of having a multitimbral synth.

Personally I think the master volume should not be able to be set over
midi, if you have one it's a global non midi editable control and then
individual channels should be controllable like most multitimbral modules
work.

**Reply by eLuSiVeMiTe** [reply]
This. 

I did try and see if it's possible to add a volume circuit/ pot to the DAC but info was sketchy and I'm not that savy with the pi's

**Reply by probonopd** [reply]
So you are saying there should be a master volume setting but it should not react to any MIDI CC?

**Reply by craigyjp** [reply]
That's my thoughts, either implemented in hardware with a stereo pot/buffer after the final output or in software if possible, but you might have to add a menu option for that and you don't want to be menu diving the master volume, doesn't make sense.

Keep it simple, I would prefer a hardware implementation and I think I will do that with mine at some point.

**Reply by now-its-dark** [reply]
Definitely think non-midi master volume parameter has merit, simply to define the max volume ceiling— afaik, different DAC configurations could have different output levels, so it would be more about defining a predictable output across devices than as a way to quickly change the volume during use.

**Reply by diyelectromusic** [reply]
> So you are saying there should be a master volume setting but it should not react to any MIDI CC?

MIDI CC 07 used to be "main volume" but then it changed to "channel volume" at some point and there is now a "universal system exclusive" message that corresponds to "master volume" instead.  So whilst that is all almost certainly after the original DX7's time, I guess that is probably the most useful way to change volume over MIDI today?

Update: I see later in this thread this started to come out, so ignore me :)

Aside: Is it worth creating a MIDI Implementation Chart equivalent on the Wiki for this kind of thing?

Ref: https://www.midi.org/specifications-old/item/table-3-control-change-messages-data-bytes-2 - CC message 07
Ref: https://www.midi.org/specifications-old/item/table-4-universal-system-exclusive-messages - SysEx message 7F-04-01

**Reply by probonopd** [reply]
Hi @diyelectromusic, yes, having a MIDI Implementation Chart for MiniDexed in the wiki would be awesome!

**Reply by diyelectromusic** [reply]
> Hi @diyelectromusic, yes, having a MIDI Implementation Chart for MiniDexed in the wiki would be awesome!

I've made a start here: https://github.com/probonopd/MiniDexed/wiki/MIDI-Implementation

It isn't in the format of "traditional" MIDI implementation charts, but I've listed everything I can find that is mentioned in mididevice.cpp.  Hopefully people can play around with it as things change and change the format if there is a better way to show it.

**Comment by eLuSiVeMiTe** [comment]
Better luck today.  So it can be added to the board. Though those pins are in use 

I'm pretty new to pi. So learning as I go. 

Bought the pi for my eldest and borrowed it for this project. 

![Screenshot_20220424-165450](https://user-images.githubusercontent.com/103902615/164985082-99837c56-b74c-4a2a-9b2d-0ffc49cd90dc.png)

**Reply by craigyjp** [reply]
Couldn't you simply put a stereo pot on the final outputs with the center of each pot as the output to the amp, 100k stereo log pot should do the trick, you could always add a pair of opamps as buffers if you really want to over engineer it.

**Comment by fp64lib** [comment]
I fully support having a master volume - in addition to the volume levels on the voices/TG. The volume levels on the voices/TG you use to come up with a balanced mix of the voices, e.g. like combining a piano sound (e.g. level 99) with some strings in the background (e.g. level 40). The master volume make the entire setting louder or softer, by keeping the balance of the mix. 

Should it be settable via MIDI? A lot of synths just add a potentiometer after all the mixing that just regulates the peak voltage of the analog output. So this is not settable. However, the Yamaha MODX has two master volume knobs, one for the analog output and a second for the USB output. There you can set the master volume with a special Sysex message, additionally to the master volume CCs which set the volume for the voices (called parts by Yamaha for the MODX/Montage).

My preference would first to have a master volume at all. Making it setable via MIDI would be a later option for me.

**Reply by probonopd** [reply]
> There you can set the master volume with a special Sysex message, additionally to the master volume CCs which set the volume for the voices (called parts by Yamaha for the MODX/Montage).

Can you find out which special Sysex they are using?

**Reply by fp64lib** [reply]
Page 89 of MODX Synthesizer Parameter Handbook, German Edition: MIDI Master Volume F0 7F 7F 04 01 ll mm F7
ll (LSM) = ignored
mm (MSB) = master volume level

**Comment by probonopd** [comment]
If someone wants to send a pull request, I'd be happy to merge it.

**Reply by dcoredump** [reply]
Found this:

[http://midi.teragonaudio.com/tech/midispec/mastrvol.htm](http://midi.teragonaudio.com/tech/midispec/mastrvol.htm)

**Comment by dcoredump** [comment]
> If someone wants to send a pull request, I'd be happy to merge it.

I will try to add SYSEX Master Volume to the SYSEX engine I am currently writing.

For my MiniDexed case I am adding a passive stereo pot for master volume.

**Reply by temporal-distortion** [reply]
> > If someone wants to send a pull request, I'd be happy to merge it.
> 
> I will try to add SYSEX Master Volume to the SYSEX engine I am currently writing.
> 
> For my MiniDexed case I am adding a passive stereo pot for master volume.



> > If someone wants to send a pull request, I'd be happy to merge it.
> 
> I will try to add SYSEX Master Volume to the SYSEX engine I am currently writing.
> 
> For my MiniDexed case I am adding a passive stereo pot for master volume.

May i dare ask how you did so and how it worked for you?

**Comment by probonopd** [comment]
MIDI 1.0 Detailed Specification says:

![image](https://user-images.githubusercontent.com/2480569/165634181-49e666d3-2ec4-43da-b404-f200e34b5fc3.png)
